cli/internal/otel/receiver.go

```diff
@@ -0,0 +1,358 @@
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/collector/receiver"
+	"go.opentelemetry.io/collector/receiver/otlpreceiver"
+	noopmetric "go.opentelemetry.io/otel/metric/noop"
+	nooptrace "go.opentelemetry.io/otel/trace/noop"
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+)
+
+type Span struct {
+	TraceID        string         `json:"trace_id"`
+	SpanID         string         `json:"span_id"`
+	ParentSpanID   string         `json:"parent_span_id"`
+	Name           string         `json:"name"`
+	StartTimestamp time.Time      `json:"start_timestamp"`
+	EndTimestamp   time.Time      `json:"end_timestamp"`
+	Attributes     map[string]any `json:"attributes"`
+}
+
+type Metric struct {
+	Name        string         `json:"name"`
+	Description string         `json:"description"`
+	Unit        string         `json:"unit"`
+	Value       int64          `json:"value"`
+	Attributes  map[string]any `json:"attributes"`
+}
+
+type TableMetric struct {
+	Table     string     `json:"table"`
+	ClientId  string     `json:"client_id"`
+	StartTime *time.Time `json:"start_time"`
+	EndTime   *time.Time `json:"end_time"`
+	Resources int64      `json:"resources"`
+	Errors    int64      `json:"errors"`
+	Panics    int64      `json:"panics"`
+}
+
+type ConsumeSpan = func(context.Context, Span)
+
+type ConsumeMetric = func(context.Context, Metric)
+
+type Consumer struct {
+	consumeSpan     ConsumeSpan
+	quit            chan any
+	metricsFilename string
+	metricsFile     *os.File
+	consumeMetric   ConsumeMetric
+}
+
+func (c *Consumer) Shutdown(ctx context.Context) error {
+	var err error
+	if c.quit != nil {
+		close(c.quit)
+	}
+	if c.metricsFile != nil {
+		err = errors.Join(err, c.metricsFile.Close())
+	}
+	return err
+}
+
+func newDefaultSpanConsumer() ConsumeSpan {
+	return func(ctx context.Context, span Span) {
+		// do nothing
+	}
+}
+
+func newDefaultMetricConsumer(metricsFile *os.File, quit chan any) ConsumeMetric {
+	tableLock := sync.Mutex{}
+	metricsMap := make(map[string]*TableMetric)
+	ticker := time.NewTicker(20 * time.Second)
+
+	renderTable := func() {
+		tableLock.Lock()
+		metrics := maps.Values(metricsMap)
+		tableLock.Unlock()
+		_, err := metricsFile.Seek(0, 0)
+		if err != nil {
+			return
+		}
+		t := table.NewWriter()
+		t.SetOutputMirror(metricsFile)
+		t.AppendHeader(table.Row{"Table", "Client ID", "Start Time", "End Time", "Resources", "Errors", "Panics"})
+		sort.SliceStable(metrics, func(i, j int) bool {
+			m1 := metrics[i]
+			m2 := metrics[j]
+
+			if m1.EndTime == nil && m2.EndTime != nil {
+				return true
+			}
+
+			if m1.EndTime != nil && m2.EndTime == nil {
+				return false
+			}
+
+			return m1.Table+m1.ClientId < m2.Table+m2.ClientId
+		})
+		for _, metrics := range metrics {
+			row := table.Row{
+				metrics.Table,
+				metrics.ClientId,
+				metrics.StartTime,
+				metrics.EndTime,
+				metrics.Resources,
+				metrics.Errors,
+				metrics.Panics,
+			}
+			if metrics.EndTime == nil {
+				row[3] = "N/A"
```

Comment:
I'm aware that `row`'s type is `[]interface{}`, but unsure how the code that uses it works. Is it ok to mix `EndTime *time.Time` with `string` on `row[3]`?

Reply:
Or so without this the table with render `<nil>` for nil values, I'll post an example for how the content of the file looks like in a bit